### DISCUSSION TITLE
ESP Now use problem

### DISCUSSION THREAD

**Discussion by glfbvr** [HomeSpan/HomeSpan#478] [Q&A]
Hi everyone !

First of all, thank you so much for the homespan lib, it's really amazing !
I'd like to use 2 ESP32 boards, one is a homespan object and added to the home app (let's call it board A), the other will receives orders from the first one (let's call it board B). Both of the boards are communicating with ESP NOW and I use a HomePod as a bridge.

I carefully read the Spanpoint examples, but didn't really manage to make it work. Unlike the example, Id' like that Board A send datas to board B, when I turn the switch on on the home app. I tried the examples, it worked well (apparently the channel used is 3), but in my case it's the same ESP32 board that is an object AND send the datas... I get a "Send Fail" error, and the log are only saying that the channel 11 was used (why aren't all the other channels tested ?)

Here is my code 

**Board B**
`
#include <WiFi.h>
#include "HomeSpan.h"

SpanPoint *chef;
float moment=0;

void setup() {
  Serial.begin(115200);
 
  Serial.print("ESP Board MAC Address:  ");
  Serial.println(WiFi.macAddress());

  chef=new SpanPoint("AA:12:32:C4:3E:1E",0,sizeof(float));
  homeSpan.setLogLevel(1);

  homeSpan.begin();

}

void loop() {
  
  homeSpan.poll();
  if(chef->get(&moment)){......};

}
`

**Board A**

`
#include <esp_now.h>
#include <WiFi.h>

#include "HomeSpan.h" 
#include "DEV_LED.h"

SpanPoint *wall;

float moment_wall=0;

void setup() {

  Serial.begin(115200);
  Serial.print("ESP Board MAC Address:  ");
  Serial.println(WiFi.macAddress());
  wall=new SpanPoint("11:E4:24:DD:B1:1D",sizeof(float),0);
  
  homeSpan.begin(Category::Lighting,"Test");
  homeSpan.setLogLevel(1);
  new SpanAccessory(1); 
  
    new Service::AccessoryInformation(); 
      new Characteristic::Identify();
      new Characteristic::Name("Test");      
                 
      etat = new DEV_LED();

          
} 

void envoie(){
  
  
    boolean success = wall->send(&moment_wall);
    
    Serial.printf("Send %s\n",success?"Succeded":"Failed");
   
}

void loop(){
  
  homeSpan.poll();
  
  //Serial.println(etat->power->getNewVal());
  if(etat->power->getVal())
  {
    moment_wall=1;
    envoie();}
    
    
delay(1);
  
  
}
`

I simplified the code. Can you help me to figure out why this doesn't work ? I tried a lot of things, without success.....

Thanks a lot for your time ! :)

**Comment by HomeSpan** [comment]
If the remote board (the one not connected to HomeKit) is only receiving data from the main board it will not be able to perform a self-calibration to match the channel.  To solve, you can either:

* Cause the remote board to send some dummy information to the main board every minute or so.  The main board does not need to do anything with this data, and you don't even need to configure SpanPoint on the main device to receive such data.  The remote board will be able to detect that it reached the main device and calibrate its channel as needed, OR

* You can "set" the channel to be used on the remote device by restricting the allowed channels on that device to just the channel you know works.

I recommend the first solution since it does not hardcode the channel number into the remote board, and if the channel on the main board changes (because your home network changes), the remote board will re-calibrate the next time it tries and fails to send dummy data.

**Comment by glfbvr** [comment]
Hi ! Thanks for the really fast answer !! I'll try the first solution, I'll keep you in touch !
Have a great day ! :)

**Comment by glfbvr** [comment]
So I configured the remote device to send a dummy message regularly (with was a success), but the main device still can't send anything... :/ 
I get this 
Send Failed
SpanPoint: Sending 4 bytes to MAC Address 11:E4:24:DD:B1:1D using channel 11...
SpanPoint: Sending 4 bytes to MAC Address 11:E4:24:DD:B1:1D using channel 11...
SpanPoint: Sending 4 bytes to MAC Address 11:E4:24:DD:B1:1D using channel 11...


but it does not perform any try on the other channels, is that normal ?

**Comment by HomeSpan** [comment]
The device attached to HomeKit cannot use any other channel except that one that is used by your WiFi network.  If your WiFi network uses Channel 11, your main device will be fixed to channel 11 so it can maintain connectivity with your WiFi network.  Remote devices are free to change their channels since they are not connecting to a WiFi network.

**Reply by glfbvr** [reply]
Thanks for your reply ! After a reset, it actually worked ! Thanks a lot for your help :)

**Comment by HomeSpan** [comment]
Super - really glad to hear it is now working.
